packages/voting/src/deposit.rs

```diff
@@ -88,16 +113,14 @@ pub fn get_deposit_msg(
             if info.deposit.is_zero() {
                 Ok(vec![])
             } else {
-                let transfer_msg = WasmMsg::Execute {
-                    contract_addr: info.token.to_string(),
-                    funds: vec![],
-                    msg: to_binary(&cw20::Cw20ExecuteMsg::TransferFrom {
-                        owner: sender.to_string(),
-                        recipient: contract.to_string(),
-                        amount: info.deposit,
-                    })?,
-                };
-                let transfer_msg: CosmosMsg = transfer_msg.into();
+                let asset = AssetBase::new(info.token.clone(), info.deposit);
+                let transfer_msg = match asset.info {
+                    AssetInfoBase::Cw20(ref _addr) => asset.transfer_from_msg(sender, contract),
+                    AssetInfoBase::Native(ref _denom) => asset.transfer_msg(contract),
```

Comment:
Another tricky one. This won't actually require the deposit to be paid if, at the time of proposal creation, there are are `info.deposit` or more tokens in the proposal module.

Tests don't catch this because they're only testing the case where a single proposal being created. In that case the contract has zero native tokens so not paying the deposit will cause it to transfer `info.deposit` tokens to itself which it does not have. If it already had a deposit inside of it that would work regardless of if the proposer paid the deposit.

Great example of why breaking this logic out and having it fail open is so important..

Reply:
Ah, really good catch.